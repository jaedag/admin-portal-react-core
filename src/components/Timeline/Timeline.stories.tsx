import React from 'react'
import { action } from '@storybook/addon-actions'
import { MemoryRouter } from 'react-router'
import Timeline, { TimelineProps, TimelineElement } from './Timeline'
import { Meta, StoryFn } from '@storybook/react'
import { Member } from '@/types'

const record: TimelineElement[] = [
  {
    historyRecord: 'Some history record',
    loggedBy: {
      firstName: 'John',
      lastName: 'Doe',
    } as Member,
    timeStamp: '2022-04-22T09:30:00.000Z',
    createdAt: {
      date: '2022-04-22T00:00:00.000Z',
    },
  },
  {
    historyRecord: 'Another history record',
    loggedBy: {
      firstName: 'Jane',
      lastName: 'Doe',
    } as Member,
    timeStamp: '2022-04-23T10:00:00.000Z',
    createdAt: {
      date: '2022-04-23T00:00:00.000Z',
    },
  },
  {
    historyRecord: 'Yet another history record',
    loggedBy: {
      firstName: 'Jane',
      lastName: 'Doe',
    } as Member,
    timeStamp: '2022-04-24T10:00:00.000Z',
    createdAt: {
      date: '2022-04-24T00:00:00.000Z',
    },
  },
]

export default {
  title: 'Components/Timeline',
  component: Timeline,
  args: { record: record, clickCard: action('clicked'), limit: 10 },
} as Meta

const Template: StoryFn<TimelineProps> = (args: TimelineProps) => (
  <MemoryRouter>
    <Timeline {...args} record={record} clickCard={action('clicked')} />
  </MemoryRouter>
)

export const Default = Template.bind({})
Default.args = {
  record: record,
  limit: 10,
}

export const Limited = Template.bind({})
Limited.args = {
  limit: 1,
}

export const ChurchModifier = Template.bind({})
ChurchModifier.args = {
  limit: 10,
  modifier: 'church',
}
