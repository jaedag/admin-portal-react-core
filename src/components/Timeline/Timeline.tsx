import React from 'react'
import { useNavigate } from 'react-router-dom'
import { Box, Flex, List, ListIcon, ListItem, Text } from '@chakra-ui/react'
import { FaCircleNotch, FaInfoCircle } from 'react-icons/fa'
import { parseDate, parseNeoTime } from '@/utils'
import { Member } from '@/types'

export type TimelineElement = {
  historyRecord: string
  loggedBy: Member
  timeStamp: string
  createdAt: {
    date: string
  }
}

export type TimelineProps = {
  limit: number
  modifier?: 'church'
  record: TimelineElement[]
  clickCard: (cardToClick: unknown) => void
}

const Timeline = (props: TimelineProps) => {
  const { record, limit, modifier, clickCard } = props

  const navigate = useNavigate()

  if (!record) {
    return null
  }

  if (record && modifier === 'church') {
    return (
      <List spacing={1}>
        {record.map(
          (element: TimelineElement, index: number) =>
            index < limit && (
              <ListItem key={index} data-testid="timeline-element">
                <Flex>
                  <Box>
                    <ListIcon as={FaInfoCircle} color="green.500" />
                  </Box>
                  <Text as="span" fontSize="sm">
                    {element.historyRecord}
                    <Text color="gray" fontSize="xs">
                      {`${parseDate(element.createdAt?.date)} at ${parseNeoTime(
                        element.timeStamp
                      )}`}
                      {element?.loggedBy && ' by'}
                      <Text
                        as="span"
                        className="font-weight-bold"
                        onClick={() => {
                          clickCard(element?.loggedBy)
                          navigate('/member/displaydetails')
                        }}
                      >
                        {element?.loggedBy &&
                          ` ${element?.loggedBy?.firstName} ${element?.loggedBy?.lastName}`}
                      </Text>
                    </Text>
                  </Text>
                </Flex>
              </ListItem>
            )
        )}
      </List>
    )
  }

  return (
    <List spacing={2}>
      {record.map(
        (element, index) =>
          index < limit && (
            <ListItem key={index} data-testid="timeline-element">
              <Flex>
                <Box>
                  <ListIcon as={FaCircleNotch} color="green.500" />
                </Box>
                <Text as="span" fontSize="sm">
                  {element.historyRecord}
                  <Text color="gray" fontSize="xs" marginBottom={1}>
                    {`${parseDate(element.createdAt?.date)} at ${parseNeoTime(
                      element.timeStamp
                    )}`}
                    {element?.loggedBy && ' by'}
                    <Text
                      as="span"
                      className="font-weight-bold"
                      onClick={() => {
                        clickCard(element?.loggedBy)
                        navigate('/member/displaydetails')
                      }}
                    >
                      {element?.loggedBy &&
                        ` ${element?.loggedBy?.firstName} ${element?.loggedBy?.lastName}`}
                    </Text>
                  </Text>
                </Text>
              </Flex>
            </ListItem>
          )
      )}
    </List>
  )
}

export default Timeline
