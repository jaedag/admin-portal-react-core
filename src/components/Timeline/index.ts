export { default, TimelineElement } from './Timeline'
