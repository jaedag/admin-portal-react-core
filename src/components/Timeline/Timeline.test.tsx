import React from 'react'
import { render } from '@testing-library/react'
import Timeline, { TimelineElement, TimelineProps } from './Timeline'
import '@testing-library/jest-dom'

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}))

const mockClickCard = jest.fn()

const mockRecord: unknown[] = [
  {
    historyRecord: 'Lorem ipsum dolor sit amet',
    loggedBy: {
      firstName: 'John',
      lastName: 'Doe',
    },
    timeStamp: '2022-04-25T20:30:00.000Z',
    createdAt: {
      date: '2022-04-25T20:30:00.000Z',
    },
  },
  {
    historyRecord: 'consectetur adipiscing elit',
    loggedBy: {
      firstName: 'Jane',
      lastName: 'Doe',
    },
    timeStamp: '2022-04-26T08:15:00.000Z',
    createdAt: {
      date: '2022-04-26T08:15:00.000Z',
    },
  },
]

describe('Timeline', () => {
  const timelineProps = {
    limit: 2,
    record: mockRecord,
    clickCard: mockClickCard,
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders the timeline correctly', () => {
    const container = render(<Timeline {...(timelineProps as TimelineProps)} />)

    expect(container).toMatchSnapshot()
  })

  describe('Timeline', () => {
    it('only renders the specified number of timeline elements', () => {
      const limit = 2
      const { getAllByTestId } = render(
        <Timeline
          record={mockRecord as TimelineElement[]}
          clickCard={() => console.log('Clicked')}
          limit={limit}
        />
      )

      const timelineElements = getAllByTestId('timeline-element')
      expect(timelineElements.length).toBe(limit)
    })
  })
})
