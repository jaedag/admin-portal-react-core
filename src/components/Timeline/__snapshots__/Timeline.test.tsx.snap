// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders the timeline correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="css-1ahh3va"
        role="list"
      >
        <li
          class="css-0"
          data-testid="timeline-element"
        >
          <div
            class="css-k008qs"
          >
            <div
              class="css-0"
            >
              <svg
                class="chakra-icon css-qld8i7"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="presentation"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
                />
              </svg>
            </div>
            <span
              class="chakra-text css-g5hezf"
            >
              Lorem ipsum dolor sit amet
              <p
                class="chakra-text css-1fzpds9"
              >
                Mon Apr 25 2022 at 20:30:00
                 by
                <span
                  class="chakra-text font-weight-bold css-0"
                >
                   John Doe
                </span>
              </p>
            </span>
          </div>
        </li>
        <li
          class="css-0"
          data-testid="timeline-element"
        >
          <div
            class="css-k008qs"
          >
            <div
              class="css-0"
            >
              <svg
                class="chakra-icon css-qld8i7"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="presentation"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
                />
              </svg>
            </div>
            <span
              class="chakra-text css-g5hezf"
            >
              consectetur adipiscing elit
              <p
                class="chakra-text css-1fzpds9"
              >
                Tue Apr 26 2022 at 08:15:00
                 by
                <span
                  class="chakra-text font-weight-bold css-0"
                >
                   Jane Doe
                </span>
              </p>
            </span>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="css-1ahh3va"
      role="list"
    >
      <li
        class="css-0"
        data-testid="timeline-element"
      >
        <div
          class="css-k008qs"
        >
          <div
            class="css-0"
          >
            <svg
              class="chakra-icon css-qld8i7"
              fill="currentColor"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
              />
            </svg>
          </div>
          <span
            class="chakra-text css-g5hezf"
          >
            Lorem ipsum dolor sit amet
            <p
              class="chakra-text css-1fzpds9"
            >
              Mon Apr 25 2022 at 20:30:00
               by
              <span
                class="chakra-text font-weight-bold css-0"
              >
                 John Doe
              </span>
            </p>
          </span>
        </div>
      </li>
      <li
        class="css-0"
        data-testid="timeline-element"
      >
        <div
          class="css-k008qs"
        >
          <div
            class="css-0"
          >
            <svg
              class="chakra-icon css-qld8i7"
              fill="currentColor"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
              />
            </svg>
          </div>
          <span
            class="chakra-text css-g5hezf"
          >
            consectetur adipiscing elit
            <p
              class="chakra-text css-1fzpds9"
            >
              Tue Apr 26 2022 at 08:15:00
               by
              <span
                class="chakra-text font-weight-bold css-0"
              >
                 Jane Doe
              </span>
            </p>
          </span>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
